File: beam-series-go/model_guide/ptransform_combine_flatten_partition.go
package main

import (
	"context"
	"log"
	"math"
  "math/rand"
	"github.com/apache/beam/sdks/v2/go/pkg/beam"
	"github.com/apache/beam/sdks/v2/go/pkg/beam/register"
	"github.com/apache/beam/sdks/v2/go/pkg/beam/x/beamx"
	"github.com/apache/beam/sdks/v2/go/pkg/beam/x/debug"
)

type Employee struct {
  Id int
  Name string 
  JoiningDate string 
  DeptID string 
  IsActive bool 
}

type LastJoiningDateFn struct {}

func (fn *LastJoiningDateFn) CreateAccumulator() string {
  return "0000-00-00"
}

func (fn *LastJoiningDateFn) AddInput(acc string, element Employee) string {
  if element.JoiningDate >= acc {
    return element.JoiningDate
  } else {
    return acc
  }
}

func (fn *LastJoiningDateFn) MergeAccumulators(acc1, acc2 string) string {
  if acc1 >= acc2 {
    return acc1
  } else {
    return acc2
  }
}

func (fn *LastJoiningDateFn) ExtractOutput(acc string) string {
  return acc
}

func max (a, v int) int {
  return int(math.Max(float64(a),float64(v)))
}

func partition_fn(emp Employee) int {
  return rand.Intn(3)
}
func init() {
  register.Function2x1(max)
  register.Function1x1(partition_fn)
}

func main() {

  employees := []Employee{
    Employee{
      0,
      "John C",
      "2000-01-01",
      "D101",
      true,
    },
    Employee{
      1,
      "Tom D",
      "2002-02-01",
      "D102",
      true,
    },
    Employee{
      2,
      "Max",
      "2003-04-01",
      "D104",
      false,
    },
    Employee{
      3,
      "Bruce",
      "2003-05-01",
      "D102",
      false,
    },
    Employee{
      4,
      "Barry",
      "2003-04-22",
      "D101",
      false,
    },
    Employee{
      5,
      "Clark",
      "2001-04-01",
      "D103",
      false,
    },
    Employee{
      6,
      "Bryan C.",
      "2010-07-01",
      "D104",
      true,
    },
  }
     
  beam.Init()

  p := beam.NewPipeline()

  s := p.Root()



  employees_pcoll := beam.CreateList(s,employees)
  employees_id := beam.ParDo(s, func(element Employee) (int) {
    return element.Id
  },employees_pcoll)

  max_id_employee := beam.Combine(s,max,employees_id)
  
  employees_kv := beam.ParDo(s, func(element Employee) (string,int) {
    return element.DeptID,element.Id
  },employees_pcoll)
  max_id_by_dept := beam.CombinePerKey(s,max,employees_kv)

  employees_kv_full := beam.ParDo(s, func(element Employee) (string,Employee) {
    return element.DeptID,element
  },employees_pcoll)

  max_joining_date_by_dept := beam.CombinePerKey(s,&LastJoiningDateFn{},employees_kv_full)
  pcoll_groups := beam.Partition(s,3,partition_fn,employees_pcoll)
  pcoll_union := beam.Flatten(s,pcoll_groups...)

   
  debug.Print(s,max_id_employee)
  debug.Print(s,max_id_by_dept)
  debug.Print(s,max_joining_date_by_dept)
  debug.Print(s,pcoll_groups[1])
  debug.Print(s,pcoll_union)

  if err := beamx.Run(context.Background(), p); err != nil {
    log.Fatalf("Failed to execute job: %v",err)
  }
}

File: beam-series-python/model_guide/ptransform_combine_flatten_partition.py
import logging
import apache_beam as beam
import random
from apache_beam.options.pipeline_options import PipelineOptions

employees = [
        {
            "id": 0,
            "name": "John C",
            "joining_date": "2000-01-01",
            "dept_id": "D101",
            "is_active": True
            },
        {
            "id": 1,
            "name": "Tom D",
            "joining_date": "2002-02-01",
            "dept_id": "D102",
            "is_active": True
            },
        {
            "id": 2,
            "name": "Max",
            "joining_date": "2003-04-01",
            "dept_id": "D104",
            "is_active": False
            },
        {
            "id": 3,
            "name": "Bruce",
            "joining_date": "2003-05-01",
            "dept_id": "D102",
            "is_active": False
            },
        {
            "id": 4,
            "name": "Barry",
            "joining_date": "2003-04-22",
            "dept_id": "D101",
            "is_active": False
            },
        {
            "id": 5,
            "name": "Clark",
            "joining_date": "2001-04-01",
            "dept_id": "D103",
            "is_active": False
            },
        {
            "id": 6,
            "name": "Bryan C.",
            "joining_date": "2010-07-01",
            "dept_id": "D104",
            "is_active": True
            }
        ]

def last_joining_date(values):
    return max(values)
def partition_fn(employee, num_partitions):
    return random.randrange(num_partitions)
class LastDateFn(beam.CombineFn):
    def create_accumulator(self):
        return "0000-00-00"
    def add_input(self,acc,input):
        return(max(acc,input["joining_date"]))
    def merge_accumulators(self,acc):
        return max(acc)
    def extract_output(self, acc):
        return acc

def run():
    pipeline_options = PipelineOptions()
    with beam.Pipeline(options=pipeline_options) as p:
        employees_pcoll = p | "Create employees pcoll" >> beam.Create(employees)
        employees_id = employees_pcoll | "Get Id" >> beam.Map(lambda x: x["id"])
        max_id_employees = employees_id | "Max ID Globally" >> beam.CombineGlobally(max)
        employees_kv = employees_pcoll | "Add keys for employees" >> beam.WithKeys(lambda x: x["dept_id"])
        employees_id_kv = employees_kv | "Extract Only ID" >> beam.Map(lambda x: (x[0],x[1]["id"]))
        max_id_per_dept = employees_id_kv | "Max ID per Dept" >> beam.CombinePerKey(max)
        employees_joining_date_kv = employees_kv | "Extract Only joining date" >> beam.Map(lambda x: (x[0],x[1]["joining_date"]))
        max_joining_date_by_dept = employees_joining_date_kv | "Max Joining Date per Dept" >> beam.CombinePerKey(last_joining_date)
        max_joining_date_by_dept_complex = employees_kv | "Max Joining Date per Dept Comples" >> beam.CombinePerKey(LastDateFn())
        
        pcoll_groups = employees_pcoll | "Partition Random" >> beam.Partition(partition_fn,3)
        pcoll_union = pcoll_groups | "Union Pcollection" >> beam.Flatten()
        


        pcoll_union | "Print debugger" >> beam.Map(print)
if __name__ == "__main__":
    logging.getLogger().setLevel(logging.INFO)
    run()
